pep-0605.rst

```diff
@@ -184,11 +197,72 @@ The majority of the proposed changes in this PEP only affect the handling of
-combination of the policies for alpha and beta releases:
+Two year cadence for stable releases
+------------------------------------
+
+With the rolling pre-freeze releases available to users that are looking to
+use leading edge versions of the reference interpreter and standard library,
+this PEP proposes that the frequency of X.Y.0 releases be adjusted to publish
+a new stable release in August every two years (starting in 2021,
+around two years after the release of Python 3.8.0).
+
+This change is arguably orthogonal to the proposed changes to the handling of
+the pre-freeze release period, but the connection is that without those
+pre-release management changes, the downsides of a two year full release cadence
+would probably outweigh the upsides, whereas the opposite is true for a 12
+month release cadence (i.e. with the pre-release management changes proposed
+in this PEP in place, the downsides of a 12 month full release cadence would
+outweigh the upsides).
+
+
+Merging of the alpha and beta phases into a "pre-freeze" phase
+--------------------------------------------------------------
+
+Rather than continuing the status quo where the pre-release alpha and beta
+phases are distinct and sequential, this PEP proposes that they instead be
+combined into a single "pre-freeze" phase with a monotonically increasing serial
+number on the releases.
+
+Rather than denoting distinct phases, the "alpha" and "beta" names would
+instead indicate whether or not the release contains breaking changes to the
+full CPython C ABI:
+
+* "alpha" releases would be "ABI breaking" releases where extension modules
+  built against the full CPython ABI in the preceding pre-release may not
+  load correctly.
+* "beta" releases would be "binary compatible" releases, where extension modules
+  built against the full CPython ABI in the preceding pre-release are expected
+  to load correctly if they weren't using any provisional C APIs, or C APIs
+  that were deprecated in the previous stable release series
+
+
+Pre-freeze phase duration and cadence
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Rather than being released monthly for a period of a few months while preparing
+a new X.Y.0 release, pre-freeze releases would instead be consistently published
+every two months.
+
+The only time this would not be the case is during the two month release
+candidate period for an upcoming X.Y.0 release (see the release candidate
+section below for more details).
+
+The pre-freeze phase would typically be expected to start 2 months after the
+preceding stable X.Y.0 release.
+
+The first pre-freeze release for any new release series will always be X.Y.0a1
+(as there is no preceding release with the same ABI version markers to judge
+binary compatibility against).
+
+Pre-freeze releases would gain an additional flag in their C ABI compatibility
+markers to avoid binary compatibility issues with the eventual stable release.
```

Comment:
I still think this is unnecessary. The whole point of the beta builds is to be ABI compatible with the next release, so why would we mark them as incompatible?

Similarly, why would we mark alpha builds as incompatible with the compatible beta builds that follow them?

The best thing we could do here is add a version check into module initialisers with the full hex version so we can fail them quickly when a specific build has crossed an alpha release without being rebuilt. Anything else is just proving that we don't believe our own rhetoric about making compatible releases :)

Reply:
The point is to make them fail at the point where we cross into X.Y.0rc1, not to make them fail on any of the alpha or beta releases - users of the beta releases are on their own in that regard.

If we were to put the "recompile the world, even if you don't actually need to" gate on every alpha release, that makes the rolling stream a PITA to support even for extension modules that aren't tinkering with struct internals, while if you don't have any technically enforced rebuild gate at all, then you risk extension modules built early in the X.Y pre-release cycle causing segfaults for people that are moving directly from the previous stable release to the X.Y.0 stable release.

Reply:
I've added a note about this to the section on ABI management changes. It's certainly possible to leave it out, it just creates the risk of painful debugging session on the *stable* X.Y.0 release that get traced back to "Oh, our dependency 'superfast-binary-operation' was affected by a CPython ABI break in X.Y.0a5, but the project hasn't published a new build since then".

Whereas if the ABI tag changes, projects can publish a new wheel file post-rc1 without even needing to publish a new release of the project (since the ABI tags are different, the filenames are different, and hence can be added to an existing release), and if they haven't published one, users will fall back to building from source (as they do today), rather than downloading a potentially incompatible binary.

Reply:
I've now also incorporated the example scenario from the Discourse thread, and made it clear that a smarter scheme that allowed the use of all binaries built since that last alpha release would also be fine (I just don't know how such a scheme could actually work, given that we don't know ahead of time which alpha release is going to be the last pre-release alpha)